Add Presto function at_timezone

svm1 opened the pull request:
Implement `at time zone` feature from Presto Java, for `TimestampWithTimezone` input. 

Example usage:
```
SELECT timestamp '2012-10-31 01:00 UTC' AT TIME ZONE 'America/Los_Angeles';
2012-10-30 18:00:00.000 America/Los_Angeles
```

majetideepak commented on velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:
```diff
@@ -120,6 +120,39 @@ class DateTimeFunctionsTest : public functions::test::FunctionBaseTest {
 
+  std::optional<TimestampWithTimezone> timestampWithTimezoneAtTimezone(
+      std::optional<int64_t> timestamp,
+      const std::optional<std::string>& timeZoneName,
```
When can the arguments be optional?

svm1 replied:
I agree, but I made them as such to match the signatures of the other functions I call, like `makeRowVector()` and `makeTimestampWithTimeZoneVector()`.

svm1 replied:
True, added tests for null input.

majetideepak commented on velox/functions/prestosql/DateTimeFunctions.h:
```diff
@@ -1394,4 +1399,41 @@ struct TimeZoneMinuteFunction : public TimestampWithTimezoneSupport<T> {
+      const arg_type<Varchar>* timeZone) {
+
+    // below statements not working in init()
+    // auto timeZoneStr = std::string_view(timeZone->data(), timeZone->size());
```
The expectation is that timeZone is a constant. Your tests make this a field.
See `TEST_F(DateTimeFunctionsTest, dateTrunc)` as an example.
You should confirm on the Java side that this argument will always be a constant.

svm1 replied:
I don't think I quite understand what you mean that my tests make it a field. My test function `timestampWithTimezoneAtTimezone()` takes in the timezone string as a constant as well. 

Checked with Presto Java and it seems that this function does only expect the timezone arg as a constant.

svm1 replied:
@majetideepak I don't think I quite understand what you mean that my tests make it a field. My test function `timestampWithTimezoneAtTimezone()` takes in the timezone string as a constant as well. 

Checked with Presto Java and it seems that this function does only expect the timezone arg as a constant.

czentgr commented on velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:
```diff
@@ -3836,3 +3869,28 @@ TEST_F(DateTimeFunctionsTest, fromUnixtimeDouble) {
+      TimestampWithTimezone(-2749482486, util::getTimeZoneID("Pacific/Fiji")),
+      timestampWithTimezoneAtTimezone(
+          -2749482486, "Atlantic/Bermuda", "Pacific/Fiji"));
+}
```
Need newline here.

svm1 replied:
Added.

czentgr reviewed (commented):
Please cleanup and use the newer timestamp with timezone implementation. This uses the older row based implementation.

svm1 commented:
Rebased and modified to use new type implementation.

czentgr commented on velox/functions/prestosql/DateTimeFunctions.h:
```diff
@@ -1495,7 +1500,40 @@ struct ToISO8601Function {
+      const core::QueryConfig& config,
+      const arg_type<TimestampWithTimezone>* /*tsWithTz*/,
+      const arg_type<Varchar>* timeZone) {
+    // below statement not working in init()
```
See https://facebookincubator.github.io/velox/develop/scalar-functions.html section Access to Session Properties and Constant Inputs.
It likely depends on how the function is called to determine whether or not the timezone is a constant. When operating on a vector of values this might become a constant e.g. imagine selecting a column 
```
select col1 AT TIME ZONE 'America/Los_Angeles'
```
where col1 is a timestamp with time zone column. I'd imagine in that case the time zone is a constant where initialize could apply.
So I think you can add it here but also need to handle it in the call in case it was not set during initialize.

svm1 replied:
Thanks for pointing this out Christian, makes sense. I noticed the example in the docs handling the possibility of a constant in initialize - modified my code here accordingly.

czentgr commented on velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:
```diff
@@ -4014,3 +4048,29 @@ TEST_F(DateTimeFunctionsTest, toISO8601Date) {
   EXPECT_EQ("-0653-07-12", toISO8601("-653-07-12"));
 }
+
+TEST_F(
```
Tests with NULL input?

svm1 commented on velox/functions/prestosql/DateTimeFunctions.h:
```diff
@@ -1495,7 +1500,40 @@ struct ToISO8601Function {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  std::string sessionTzName = "";
+  const date::time_zone* sessionTimezone = nullptr;
```
Removed.

svm1 commented:
@pedroerp would appreciate if you could take a look!

pedroerp commented on velox/functions/prestosql/DateTimeFunctions.h:
```diff
@@ -1499,4 +1504,39 @@ struct ToISO8601Function {
 };
 
+template <typename T>
+struct TimestampWithTimezoneAtTimezoneFunction
```
why not simply calling it `AtTimezoneFunction` to follow the pattern from other functions?

svm1 replied:
Thanks for the review. This naming was leftover from when I had implementations for different function signatures, I glossed over it when refactoring. Cleaned them all up.

pedroerp commented on velox/functions/prestosql/DateTimeFunctions.h:
```diff
@@ -1499,4 +1504,39 @@ struct ToISO8601Function {
+      const arg_type<Varchar>* timezone) {
+    if (timezone != nullptr) {
+      targetTimezoneID_ = util::getTimeZoneID(
+          std::string_view(timezone->data(), timezone->size()));
```
nit: I think implicit conversion from StirngView into std::string_view is allowed, IIRC. Can't you just use `*timezone` here? (or `(std::string_view)*timezone`)

svm1 replied:
Thanks, explicit cast did work. Replaced.

pedroerp commented on velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:
```diff
@@ -121,6 +121,40 @@ class DateTimeFunctionsTest : public functions::test::FunctionBaseTest {
+      const std::optional<std::string>& timeZoneName,
+      const std::optional<std::string>& newTimeZoneName) {
+    facebook::velox::VectorPtr resultVector;
+    if (!timestamp.has_value() || !timeZoneName.has_value()) {
```
why can't we use `evaluateOnce()` like we do for the other functions?

svm1 replied:
Using the uncasted vector returned from `evaluate()` seemed the best way to simulate returning a `TimestampWithTimezone` "tuple". I think `evaluateOnce()` can only return primitive types.

pedroerp replied:
can't you do something like? `evaluateOnce<TimestampWithTimezone>()`

I haven't tried it myself, but I suppose this should work.

svm1 replied:
Tested that, it doesn't work. I based this implementation off `parseDatetime()`, the only other function in here to return a TimestampWithTimezone - so this likely is a limitation with `evaluateOnce()`.

pedroerp commented on velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:
```diff
@@ -4014,3 +4048,43 @@ TEST_F(DateTimeFunctionsTest, toISO8601Date) {
+    timestampAtTimezoneTestTimestampWithTimezoneInput) {
+  EXPECT_EQ(
+      TimestampWithTimezone(1500101514, util::getTimeZoneID("America/Boise")),
+      timestampWithTimezoneAtTimezone(
```
call the function `atTimezone()` to follow the pattern used for other functions?

svm1 replied:
Renamed.

pedroerp reviewed (commented):
Thank you! One last comment abvoe

pedroerp reviewed (approved):
Thank you @svm1 . We should probably still try to fix/extend `evaluateOnce()`, but this may be a bit hairier and better done in a separate PR.

pedroerp commented on velox/functions/prestosql/DateTimeFunctions.h:
```diff
@@ -1500,4 +1505,36 @@ struct ToISO8601Function {
+    }
+  }
+
+  FOLLY_ALWAYS_INLINE void call(
```
@svm1: @mbasmanova just brought up a good point that the Presto java implementation seems a bit more involved (com/facebook/presto/operator/scalar/DateTimeFunctions.java) 

could you check why that is the case? Looks like the might be something not correctly handled in this simpler implementation.

svm1 replied:
@pedroerp posting Presto implementation here for reference:

```
@ScalarFunction(value = "at_timezone", visibility = HIDDEN)
    @LiteralParameters("x")
    @SqlType(StandardTypes.TIMESTAMP_WITH_TIME_ZONE)
    public static long timestampAtTimeZone(@SqlType(StandardTypes.TIMESTAMP_WITH_TIME_ZONE) long timestampWithTimeZone, @SqlType("varchar(x)") Slice zoneId)
    {
        try {
            return packDateTimeWithZone(unpackMillisUtc(timestampWithTimeZone), zoneId.toStringUtf8());
        }
        catch (NotSupportedException | TimeZoneNotSupportedException e) {
            throw new PrestoException(NOT_SUPPORTED, e.getMessage(), e);
        }
        catch (IllegalArgumentException e) {
            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, e.getMessage(), e);
        }
    }
```

`packDateTimeWithZone()` as follows:
```
public static long packDateTimeWithZone(long millisUtc, TimeZoneKey timeZoneKey)
    {
        requireNonNull(timeZoneKey, "timeZoneKey is null");
        return pack(millisUtc, timeZoneKey.getKey());
    }
```

Seems to be doing pretty much the same thing, bitpacking the millisecond and timezone ID values. 

As for the exception handling - the equivalent of the `TimeZoneNotSupportedException` is handled implicitly here by the call to `getTimeZoneID()` which would throw a similar error if given an invalid timezone.  

The `IllegalArgumentException` should implicitly be handled by the coordinator if I'm not mistaken - as in any attempt to call the function on invalid/mismatched input should be caught before coming to this level?

Now it is interesting that all the Presto `at_timezone()` implementations take in either `time with time zone` or `timestamp with time zone` - none that take in a mere `timestamp`, as I would expect. Perhaps the values are casted prior to hitting the function, paired with session time zone?

mbasmanova replied:
@svm1 I wonder if you know what this code is for:

```
    private static long timeAtTimeZone(SqlFunctionProperties properties, long timeWithTimeZone, TimeZoneKey timeZoneKey)
    {
        DateTimeZone sourceTimeZone = getDateTimeZone(unpackZoneKey(timeWithTimeZone));
        DateTimeZone targetTimeZone = getDateTimeZone(timeZoneKey);
        long millis = unpackMillisUtc(timeWithTimeZone);

        // STEP 1. Calculate source UTC millis in session start
        millis += valueToSessionTimeZoneOffsetDiff(properties.getSessionStartTime(), sourceTimeZone);

        // STEP 2. Calculate target UTC millis in 1970
        millis -= valueToSessionTimeZoneOffsetDiff(properties.getSessionStartTime(), targetTimeZone);

        // STEP 3. Make sure that value + offset is in 0 - 23:59:59.999
        long localMillis = millis + targetTimeZone.getOffset(0);
        // Loops up to 2 times in total
        while (localMillis > TimeUnit.DAYS.toMillis(1)) {
            millis -= TimeUnit.DAYS.toMillis(1);
            localMillis -= TimeUnit.DAYS.toMillis(1);
        }
        while (localMillis < 0) {
            millis += TimeUnit.DAYS.toMillis(1);
            localMillis += TimeUnit.DAYS.toMillis(1);
        }

        try {
            return packDateTimeWithZone(millis, timeZoneKey);
        }
        catch (NotSupportedException | TimeZoneNotSupportedException e) {
            throw new PrestoException(NOT_SUPPORTED, e.getMessage(), e);
        }
        catch (IllegalArgumentException e) {
            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, e.getMessage(), e);
        }
    }
```

svm1 replied:
@mbasmanova @pedroerp To be entirely honest, this block is throwing me for a loop. 

The function `valueToSessionTimeZoneOffsetDiff()`, which is calculating a given time zone's difference in offsets between the first moment of the epoch (1970-01-01) and the session start time, has some comments explaining its purpose and the greater logic of this process, but I'm still not clear on what exactly is being achieved. 

`millis` is already representing a point in UTC, so I don't see why the function doesn't simply begin from "Step 3" where the target time zone's offset is added to it, to convert to the target time zone. 

But I do see this seems specific to Presto's `time with time zone` type and the limitations around it - namely that lacking a date component makes it more ambiguous to pin down the time zone the input value should be associated with, seems that all this logic is to circumvent that issue, if I'm not mistaken. I don't think this necessarily applies to our situation in Velox - we don't have a `time with time zone` type at all, and never face this ambiguity with `timestamp with time zone` input which is clearly demarcated by its own "starting" time zone.

mbasmanova replied:
@svm1 Thank you for taking a look. I was puzzled by this as well. As you pointed out this logic applies to TIME with time zone, not TIMESTAMP with time zone. I missed that part. Thank you for clarifying.

> lacking a date component makes it more ambiguous to pin down the time zone the input value should be associated with, seems that all this logic is to circumvent that issue

Sounds plausible.

pedroerp commented:
@svm1 looks like a bunch of builds are failing.

svm1 commented:
Rebased as my branch was a bit out of date, should take care of the build issues.

mbasmanova commented on velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:
```diff
@@ -121,6 +121,40 @@ class DateTimeFunctionsTest : public functions::test::FunctionBaseTest {
+                   TIMESTAMP_WITH_TIME_ZONE(), 1, pool()),
+               makeNullableFlatVector<std::string>({newTimeZoneName})}));
+    } else {
+      resultVector = evaluate(
```
```
evaluateOnce<int64_t>(
   "at_timezone(c0, c1)", 
   {timestamp, timeZoneName}, 
   {TIMESTAMP_WITH_TIME_ZONE(), VARCHAR()})
```

should work. Does it not work?

svm1 replied:
Setting the return type to `int64_t` mitigates one issue, but that still doesn't work because `evaluateOnce()` cannot take in a `timestamp with time zone` type.

The equivalent approach requires using `evaluateWithTimestampWithTimezone()`, which calls `evaluate()` under the hood. But this function is not equipped to take in any additional arguments with a `timestamp with time zone`, which is why I made my custom function.

Though now that I think about it, I should probably just extend the existing `evaluateWithTimestampWithTimezone()` to perhaps have it take in additional optional arguments.

mbasmanova replied:
There is this version of evaluateOnce, which on the surface is supposed to work, but .... Args is a single template parameter, not variadic like in the other version of evaluateOnce, hence, only works if there is a single argument or all arguments are of the same type, which is not our case.

```
  template <typename ReturnType, typename Args>
  std::optional<ReturnType> evaluateOnce(
      const std::string& expr,
      const std::vector<std::optional<Args>>& args,
      const std::vector<TypePtr>& types,
      const std::optional<SelectivityVector>& rows = std::nullopt,
      const TypePtr& resultType = nullptr) {
    std::vector<VectorPtr> flatVectors;
    for (vector_size_t i = 0; i < args.size(); ++i) {
      flatVectors.emplace_back(makeNullableFlatVector(
          std::vector<std::optional<Args>>{args[i]}, types[i]));
    }
    auto rowVectorPtr = makeRowVector(flatVectors);
    return evaluateOnce<ReturnType>(expr, rowVectorPtr, rows, resultType);
  }
```

I assume the following works:

```
evaluateOnce<int64_t>(
   "at_timezone(c0, c1)", 
   {ts1, ts2}, 
   {TIMESTAMP_WITH_TIME_ZONE(), TIMESTAMP_WITH_TIME_ZONE()})
```

mbasmanova replied:
We can introduce a version of evaluateOnce that takes variadic number of arguments + a list of types:

```
  template <typename ReturnType, typename... Args>
  std::optional<ReturnType> evaluateOnce(
      const std::string& expr,
      const std::vector<TypePtr>& types,
      const std::optional<Args>&... args) {
    std::vector<VectorPtr> flatVectors;

    int i = 0;
    (
        [&] {
          flatVectors.emplace_back(makeNullableFlatVector(
              std::vector<std::optional<Args>>{args}, types[i]));
          ++i;
        }(),
        ...);

    return evaluateOnce<ReturnType>(expr, makeRowVector(flatVectors));
  }
```

Then, use it like

```
evaluateOnce<int64_t>(
        "at_timezone(c0, c1)", 
        {TIMESTAMP_WITH_TIME_ZONE(), VARCHAR()},
        timestamp,
        timeZoneName);
```

svm1 replied:
@pedroerp @mbasmanova I rewrote the tests here, using the new `evaluateOnce()` framework. Certainly nice to avoid having to deal with the vectors directly, thank you Pedro! 

Please let me know if everything looks good here.

pedroerp reviewed (approved):
Thank you! Sorry about the delay, just coming back from PTO

pedroerp reviewed (changes requested):
Actually, looks like there is a bunch on build failures. Could you fix these first?

svm1 commented:
@pedroerp no worries, thanks for all your feedback! 

Fixed the build errors, had to update my `initialize()` method - and added docs as I realize I'd forgotten that.

pedroerp reviewed (approved):
Thank you!
